packit/utils/obs_helper.py

```diff
@@ -0,0 +1,453 @@
+            repo_elem.append(path_elem)
+
+        for arch in repo.arch:
+            (arch_elem := ET.Element("arch")).text = arch
+            repo_elem.append(arch_elem)
+
+        root.append(repo_elem)
+
+    return root
+
+
+def parse_changelog_subject(line: str):
+    date: list[str]
+    author: list[str] = []
+    email: str = ""
+    version: str = ""
+
+    toks = [tok.strip() for tok in line.split(" ") if tok.strip()]
+
+    toks.pop(0)  # Remove leading "*"
+    date = toks[:4]
+
+    del toks[:4]
+    for index, element in enumerate(toks):
+        if not element.startswith("<"):
+            author.append(element)
+        else:
+            email += element
+            del toks[: index + 1]
+            break
+
+    if toks and toks[0] == "-":
+        toks.pop(0)
+
+    if toks:
+        version = toks[0]
+
+    return date, author, email, version
+
+
+def format_changelog(changelog_str: str) -> str:
+    """Converts spec changelog to OBS .changes format"""
+    output: list[str] = []
+    changelog_time = "12:00:00 UTC"
+    changelog = changelog_str.splitlines()
+
+    # Remove leading newline
+    if changelog[0] == "\n":
+        changelog.pop(0)
+
+    for line in changelog:
+        if line.strip().startswith("*"):
+            date, author, email, version = parse_changelog_subject(line)
+
+            date.insert(3, changelog_time)
+            output.append("-" * 68)
+            output.append(f"{' '.join(date)} - {' '.join(author)} {email}")
+            if version:
+                output.append(f"- {version}")
+        else:
+            if re.match(r"-\s[0-9]+[\.\-]*", line):  # Match version line
+                output.append(line)
+            else:
+                output.append(f"  {line}")
+
+    return "\n".join(output)
+
+
+def create_changes_file(package_dir: Path) -> None:
+    """
+    Creates a changelog file from a by copying the changelog section from the spec file.
+    Sets release to 0 since obs handles release numbers
+
+    Args:
+        package_dir (Path): Path to the directory containing the package spec file.
+
+    Returns:
+        None
+    """
+    specfile: str
+    content: str
+    release_line = "Release:    0"
+
+    for file in list(filter(os.path.isfile, os.listdir(package_dir))):
+        if file.endswith(".spec"):
+            specfile = file
+            break
+
+        if file.endswith(".changes"):
+            logger.info(".changes file already exists in package")
+            return
+
+    if not specfile:
+        raise ValueError("Cannot find spec file in package")
+
+    with open(specfile) as s_file:
+        content = s_file.read()
+
+    split_content = re.split(r"%changelog", content, flags=re.IGNORECASE)
+    if len(split_content) < 2:
+        logger.info("Project has no changelog")
+        return
+
+    specfile_content, changes_file_content = split_content[0], "".join(
+        split_content[1:],
+    )
```

Comment:
What about using [our specfile library](https://github.com/packit/specfile/) for this?
https://packit.dev/specfile/api/specfile/changelog

It might take care of various corner cases...

Reply:
Creating follow-up: https://github.com/packit/packit/issues/2399